Complex value-type

dondieselkopf opened the pull request:
This is a first working version for complex value types. Works for GMRES with Gauss-Seidel at the moment. Other solver/relaxation algorithms should be checked and tweaked similarly.

ddemidov commented on amgcl/backend/builtin.hpp:
```diff
@@ -605,6 +606,44 @@ struct inner_product_impl<
+                typename boost::is_complex<typename value_type<Vec1>::type>::type
+                >::type
+        >::type
+    >
```
Should this specialization be moved into amgcl/backend/enable_complex.hpp?

ddemidov commented on amgcl/backend/builtin.hpp:
```diff
@@ -605,6 +606,44 @@ struct inner_product_impl<
+
+        {
+            size_t chunk_start = 0;
+            size_t chunk_end   = n;
```
This is leftover code from openmp implementation of the generic inner product. We can either remove the code and keep the loop below serial, or we can do something like
```cpp
#pragma omp parallel
{
    ...
    for(...) {
        ...
    }
#pragma omp atomic
    sum += s;
}
return sum;
```

ddemidov commented on amgcl/backend/interface.hpp:
```diff
@@ -72,6 +72,18 @@ struct value_type {
 
+/// Metafunction that returns the real value type of a compound type.
+template <class T, class Enable = void>
+struct real_type {
```
I think a better name for this would be `scalar_of`

ddemidov commented on amgcl/backend/interface.hpp:
```diff
@@ -72,6 +72,18 @@ struct value_type {
+template <class T>
+struct real_type< std::complex<T> > {
+    typedef T type;
+};
```
This should probably be moved into amgcl/backend/enable_complex.hpp

ddemidov commented on amgcl/coarsening/plain_aggregates.hpp:
```diff
@@ -61,132 +61,134 @@ namespace coarsening {
  * \ingroup aggregates
  */
 struct plain_aggregates {
-    /// Aggregation parameters.
```
Can you please use spaces for indentation here to keep the diff minimal (and to keep the file consistent with the rest of the code)?

ddemidov commented on tests/test_complex_adapter.cpp:
```diff
@@ -0,0 +1,72 @@
+        ;
+}
+
+BOOST_AUTO_TEST_SUITE_END()
```
I think it would be a good idea to keep all the complex tests in a single file (test_complex.cpp)

ddemidov commented:
Thanks! I left some comments in the code above.

ddemidov commented on amgcl/backend/builtin.hpp:
```diff
@@ -605,6 +606,44 @@ struct inner_product_impl<
+            V s = 0;
+            V c = 0;
+            for(size_t i = chunk_start; i < chunk_end; ++i) {
+                V d = x[i] * std::conj(y[i]) - c;
```
We could also introduce `conj_transp` operation that would be a noop by default, `std::conj` for complex numbers, and actual conjugate transpose for generic nonscalar values in the future. This way we could keep single implementation of inner product.

dondieselkopf commented:
Thanks for your review! I have implemented your suggested improvements. Please take a look at it if you have time. A few things may require some discussion:
1. I have implemented the inner product using conj_transp. For a complex type, complex conjugate would be sufficient. In foresitght of matrix-valued types, however, this may be good idea.
2. Do you think we need to use conj_transp in vmul_impl, too?
3. In a previous commit I changed the example matrix to have a complex diagonal. However, the ERF solver does not seem to converge with such a matrix. We should take a closer look at it.

dondieselkopf commented:
I have created the functions `is_zero`, `make_one` and `make_zero` in the interface. Do you think it makes sense?

ddemidov commented:
> Do you think we need to use `conj_transp` in `vmul_impl`, too?

I don't think so. `vmul` operation is used in relaxation methods, such as `damped_jacobi`, to express operation of [`x += w * D * y`](https://github.com/ddemidov/amgcl/blob/fd735d4ecd7e390fd0c367b88b622063a834e07b/amgcl/relaxation/damped_jacobi.hpp#L131). Here `x` and `y` are vectors (same as solution or RHS), and `D` is the inverted matrix diagonal.

> In a previous commit I changed the example matrix to have a complex diagonal. However, the ERF solver does not seem to converge with such a matrix. We should take a closer look at it.

I tried to reproduce this on top of your branch, but both complex tests pass now. Is it possible the problem has been fixed by one of your more recent commits? By the way, I simply forgot to update diagonal value in the sample matrix, so thanks for catching this!

> I have created the functions is_zero, make_one and make_zero in the interface. Do you think it makes sense?

I think that's a good idea. Did you forget to add a helper function for `make_one` though? `make_zero` is currently called as `make_zero<T>()` and delegates the work to `make_zero_impl<T>::get()`. `make_one` lacks this syntactic helper though and is called as `make_one<T>::get()`.

Also, `tests/sample_problem.hpp` needs to include `<amgcl/backend/interface.hpp>` to avoid compilation errors.

Another thing I noticed about `make_one` is that it almost exclusively is used in expressions as `make_one<T>() / v`. What do you think about adding a dedicated `inv(v)` function that would do an inversion? It could potentially be more efficient than `1/v` for matrix types.

I am not sure that `amgcl::backend` namespace is a good place for functions like `is_zero`, `conj_transpose`, `make_zero`, etc. A backend in amgcl is a set of vector and matrix operations for specific vector and matrix types (for example, located on a GPU). Do you think we should introduce `amgcl::math` or `amgcl::linalg` namespace for the above functions? Placing them into `amgcl/backend/interface.hpp` is like making a promise we will support them for some backends other than builtin, but I don't think its possible.


Otherwise, this looks good! Can you please squash the commits into a single one or may be several major steps? This can be easily done with `git rebase -i origin/master`.

ddemidov commented:
It seems you fixed the `sample_problem.hpp` while I was writing this :).

dondieselkopf commented:
Your suggestions sound good to me. Will do this asap. Yes, we should maybe introduce a new namespace. Didn't like to put this into the interface anyway. Will do something about that. Also, I can rebase the commits once we have a proper version.

dondieselkopf commented:
* I have removed the conj_transp in vmul()
* Put the complex diagonal value back, now that the test is passed by both approaches
* Put the helper functions into the namespace `amgcl::math` in file `util.hpp` (there may be a better option)
* Created delegate structs for each function
* Introduced an `inverse()` function

The rebasing somehow failed when I tried to push the changes to the remote. Will have a look at this later again, given up for the moment.

Let me know what you think.

ddemidov commented on amgcl/backend/enable_complex.hpp:
```diff
@@ -0,0 +1,99 @@
+#ifndef ENABLE_COMPLEX_HPP_
+#define ENABLE_COMPLEX_HPP_
+
```
Can you please put a copyright header here (with your name)?

ddemidov commented:
> The rebasing somehow failed when I tried to push the changes to the remote.

Did you use the force? (`git push -f`)

ddemidov commented:
> Put the helper functions into the namespace amgcl::math in file util.hpp (there may be a better option)

I think this is good for now. We can introduce new file when/if we need more functions later.

dondieselkopf commented:
Looks like the force option did the trick... Gotta read up on git basics next :)

ddemidov commented:
This looks great, thanks for working on this! Do you have plans on updating the rest of the algorithms, or is it enough for your needs?

dondieselkopf commented:
Thanks for the reviews and suggestions! Most of the other algorithms seem a bit more tricky to update. I will look at it but I cannot make any promises on a timeline. At the moment I am fine with GMRES. Although I would like to move to BiCGSTAB soon. Will let you know of any progress.
